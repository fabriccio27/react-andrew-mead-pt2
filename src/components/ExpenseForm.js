import React from "react";
import moment from "moment";
import {SingleDatePicker} from "react-dates";
import 'react-dates/initialize';
import 'react-dates/lib/css/_datepicker.css';



/* const now = moment();
console.log(now.format("Do MMM YYYY"));  //estos codigos se ven en la doc --> 28th May 2021 */
class ExpenseForm extends React.Component {
    //aca voy a usar la sintaxis vieja para poder acceder a los props y poblar el state
    constructor(props){
        super(props);
        this.state = {
            description: props.expense? props.expense.description:"",
            note: props.expense? props.expense.note:"",
            amount: props.expense?(props.expense.amount/100).toString():"",
            createdAt: props.expense?moment(props.expense.createdAt): moment(),
            calendarFocused:false,
            error:""
        };
    }
    onDescriptionChange=(event)=>{
        const description = event.target.value;
        this.setState(()=>({description}))
    };
    onNoteChange = (event) => {
        const note = event.target.value;
        this.setState(()=>({note}));
    };
    onAmountChange = (event) => {
        const amount = event.target.value;
        // para esto tengo que verificar entrada
        //hago un test de regular expression 
        const testRE = /^\d{1,}(\.\d{0,2})?$/
        if (!amount || amount.match(testRE)){
            this.setState(() => ({amount}));
        }
    };
    onDateChange = (createdAt) => {
        //la referencia dice que el argumento del handler es date, si le puse createdAt en mi state, me ahorro caracteres
        if (createdAt){
            this.setState(()=>({createdAt}));
        }
    };
    onFocusChange = ({focused}) =>{
        this.setState(()=>{
            return {calendarFocused:focused}
        })
    };
    onSubmit = (event) =>{
        event.preventDefault();
        if(!this.state.description || !this.state.amount){
            this.setState(()=>({error:"Please fill correctly both description and amount fields."}));
        }else{
            this.setState(()=>({error:""}));
            //aca no pongo el dispatch de addExpense, porque la idea es que este ExpenseForm sea reutilizable para edit
            //lo que pongo aca es una funcion que le paso a ExpenseForm como prop. En AddExpensePage esta definida.
            this.props.onSubmit({
                description:this.state.description,
                amount:parseFloat(this.state.amount,10)*100,
                note:this.state.note, 
                createdAt: this.state.createdAt.valueOf() //esto me pasa de string de fecha a timestamp, metodo de moment
            });
        }
    };
    render(){
        return(
            <div>
                ExpenseForm
                {this.state.error && <h3>{this.state.error}</h3>}
                <form onSubmit={this.onSubmit}>
                    <input 
                        type="text" 
                        autoFocus 
                        placeholder="Description" 
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input type="text" placeholder="Amount" value={this.state.amount} onChange={this.onAmountChange}/> {/* si para amount dejo type number, no puedo controlar los decimales, por eso cambio a text */}
                    <textarea placeholder="Add a note to your expense (optional)" onChange={this.onNoteChange}></textarea>
                    <SingleDatePicker
                        date={this.state.createdAt} // esta es la fecha que se va a mostrar inicialmente. En state sale de moment()
                        onDateChange={this.onDateChange} // handler para cuando user cambia fecha
                        focused={this.state.calendarFocused} // P
                        onFocusChange={this.onFocusChange} // handler para cuando react-dates necesita cambiar el valor de focused
                        numberOfMonths={1} // me muestra de a un mes en el picker
                        isOutsideRange={()=>false}
                    />
                    <button>Add Expense</button>
                </form>
            </div>
        )
    }
}

export default ExpenseForm;