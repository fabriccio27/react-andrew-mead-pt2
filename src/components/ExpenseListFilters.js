import React from "react";
import {connect} from "react-redux"; // como lo que quiero es modificar el store, necesito conectarme
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from "../actions/filters";
import {DateRangePicker} from "react-dates";

class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused:null
    };   
    onDatesChange =({startDate,endDate})=>{
        this.props.dispatch(setStartDate(startDate));
        this.props.dispatch(setEndDate(endDate));
    };
    onFocusChange=(calendarFocused)=>{
        this.setState(()=>({calendarFocused}));
    };
    render(){
        return(
            <div>
                <input type="text" value={this.props.filters.text} onChange={(event)=>{
                    props.dispatch(setTextFilter(event.target.value))
                }} />
                {/* cada vez que modifique el input, por el onChange, voy a resetear filter.text en el store
                como en ExpenseList estoy usando selectExpenses, me va cambiando que es lo que renderiza */}
                <select value={this.props.filters.sortBy} onChange={(event)=>{
                    //por como defini el reducer de filter, el default en filters.sortBy es date
                    if (event.target.value === "date") this.props.dispatch(sortByDate());
                    else if (event.target.value === "amount") this.props.dispatch(sortByAmount());
                }}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={()=>false}
                    showClearDates={true} //esto es para borrar las fechas
                    
                />
            </div>
        );
   };
}

// esto tiene que despachar una accion para setear el filter.text. 
// cuando conecto un componente, adquiero .dispatch en sus props
const mapStateToProps = (state) => {
    //aca va quÃ© parte quiero del state
    return {
        filters:state.filter
    };
};
/* const mapDispatchToProps = ()=> {

}
 */
export default connect(mapStateToProps)(ExpenseListFilters);