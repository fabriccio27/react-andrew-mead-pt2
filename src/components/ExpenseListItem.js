import React from "react";
import {Link} from "react-router-dom";


//aca en vez de usar props, podria usar un destructurado {dispatch(si lo requeria), id, description, amount, createdAt}
const ExpListItem = (props) => {
    return (
        <div>
            <h4><Link to={`/edit/${props.id}`}>{props.description}</Link></h4>
            <p>Amount : {props.amount} | Created:{props.createdAt}</p>
        </div>
    );
};


export default ExpListItem;