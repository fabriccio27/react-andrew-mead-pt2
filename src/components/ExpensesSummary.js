//tengo que obtener los expenses que cumplen con los filters
//tengo que obtener esas expenses del state como props en mi componente
//entonces voy a necesitar conexion
import React from "react";
import selectExpenses from "../selectors/expenses";
import getExpensesTotal from "../selectors/expenses-total";
import {connect} from "react-redux";
import numeral from "numeral";

const ExpensesSummary = (props) => {
    return(
        <div>
            {props.expensesCount>0? (
                <p>Total amount: {numeral(props.expensesTotal/100).format("$0,0.00")} for {props.expensesCount} expense(s)</p>
            ): (
                <p>No expenses selected</p>
            )}
        </div>
    )
};

const mapStateToProps = (state)=>{
    const selected =  selectExpenses(state.expenses, state.filter);
    return {
        expensesTotal:getExpensesTotal(selected),
        expensesCount:selected.length
    };
}
export {ExpensesSummary};
export default connect(mapStateToProps)(ExpensesSummary);
    
