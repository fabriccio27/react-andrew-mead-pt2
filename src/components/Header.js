import React from "react";
import {NavLink} from "react-router-dom";

const Header = () => (
    <div>
        <h1>expensify</h1>
        <NavLink exact={true} to="/" activeClassName="is-active">Dashboard</NavLink>
        <NavLink to="/create" activeClassName="is-active">Create Expense</NavLink>
        {/* <NavLink to="/edit" activeClassName="is-active">Edit Expense</NavLink>  este voy a acceder desde un elem en particular*/}
        {/* <NavLink to="/help" activeClassName="is-active">Help</NavLink> */}
    </div>
)

export default Header;