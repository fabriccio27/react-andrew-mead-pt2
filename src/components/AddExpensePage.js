import React from "react";
import ExpenseForm from "./ExpenseForm";
import {connect} from "react-redux";
import {addExpense} from "../actions/expenses"

//lo exporto sin conectar para poder testearlo
export class AddExpensePage extends React.Component {
    onSubmit=(expense)=>{
        this.props.onSubmit(expense);
        this.props.history.push("/");
    };
    render(){
        return(
            <div>
                <h2>Add Expense</h2>
                <ExpenseForm onSubmit={this.onSubmit} /> 
                {/* al no pasar funciones inline, no tiene que calcularlas en cada render */}
                {/* en ExpenseForm le estoy pasando un objeto a this.onSubmit */}
            </div>
        );
    };
}


const mapDispatchToProps = (dispatch) =>{
    return {
        onSubmit:(expense)=> dispatch(addExpense(expense))
    };
};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);