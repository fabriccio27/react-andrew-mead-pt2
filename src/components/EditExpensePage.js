import React from "react";
import ExpenseForm from "./ExpenseForm";
import {connect} from "react-redux"; //quiero modificar el state a traves de esta pagina, asi que necesito conectar
import {editExpense, startRemoveExpense} from "../actions/expenses";
//lo que quiero es despachar la accion que me permita borrar segun id, una expensa
// - tengo que conectar componente al store, pero si no requiero nada del state (porque ya obtuve de props en este caso), dejo () vacios


export class EditExpensePage extends React.Component {
    onSubmit = (expense) =>{
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push("/");
    };
    onClick = () =>{
        this.props.startRemoveExpense(this.props.expense.id); //el objeto seria data, segun lo que puse en mapDispatchToProps
        this.props.history.push("/");
    };
    render(){
        return(
            <div>
                {/* This is the edit expense page for {this.props.match.params.id} */}
                <ExpenseForm
                    expense={this.props.expense} /* esto pasa a expense form y tiene que poblar el state, en lo posible */
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onClick}>Remove</button>
            </div>
        );
    };
}

//literamente es sacar de state y pasar como props, a quien? al componente a conectar
const mapStateToProps = (state, props) =>{
    // quiero de state.expenses, la expense que tenga el mismo id que el id que saco de props
    //el metodo .find() para arrays es como un filter, pero devuelve un solo match, el 1ro
    return {
        expense: state.expenses.find((expense)=>{
            return expense.id === props.match.params.id;
        })
    };
};

const mapDispatchToProps = (dispatch) =>{
    /* a las props que accedo en las funciones, le mapoe una funcion que incluye dispatch 
    y una accion, pasar los parametros bien */
    return {
        editExpense:(id,expense) => dispatch(editExpense(id,expense)),
        startRemoveExpense:(data) => dispatch(startRemoveExpense(data))
    };
}


export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);

// en history tengo lo que necesita react-router para hacer client-side routing, metodos para redirigir y demas solo con js
// componentes que no fueron seteados para un Route no tienen props --> Header no recibe props. 
// en location tengo info de query strings si las hubiera, y de hash
// en match tengo info de parametros URL, por ejemplo localhost/edit/45, 45 es un para URL, el cual puede ser un id p/una db
    //para poder agregar esos params URL tengo que configurar el path de edit en el componente Route a path="/edit/:id"
    //esta disponible en match.params como para {"id":4556}, por ejemplo