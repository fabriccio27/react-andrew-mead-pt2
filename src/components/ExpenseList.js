import React from "react";
import {connect} from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

//IMPORTANTE cuando conecto componente al store de Redux, se vuelve reactivo, se actualiza ante el cambio de store
//no necesito usar store.subscribe o store.getState...

//exporto para el testeo
export const ExpenseList = (props) =>(
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-larger">Expense</div>
            <div className="show-for-larger">Amount</div>
        </div>
        <div className="list-body">
        {
            props.expenses.length === 0? (
                <div className="list-item list-item--message">
                    <span>No expenses to show</span>
                </div>
            ) : (
                props.expenses.map((expense)=>{
                    return <ExpenseListItem key={expense.id} {...expense} />
                })
            )
        }
        </div>
    </div>
);

//en mi ejemplo de HOC, yo creaba la funcion, aca la función es el return value de connect(...) (sería mi withAdminWarning)
//llamo a connect con que parte del estado quiero acceso, uso una callback
//en general no creo este ConnectComponent, lo exporto anonimamente, esto de abajo fue para ilustrar
/* 
const ConnectedExpenseList = connect((state)=>{
    return {
        //puedo poner cualquier key:value que yo quiero, gralmente del state, pero puedo pasar cualquier cosa (x ej, name:"Fabricio")
        expenses:state.expenses
    }; //esto va a estar disponible como props al componente conectado
})(ExpenseList);
export default ConnectedExpenseList; */

/* otro pattern mas resumido
export default connect(state=>{
    return {
        expenses:state.expenses
    };
})(ExpenseList);
*/
// el que ya conocia es crear mapStateToProps
const mapStateToProps = (state)=>{
    return {
        //expenses:state.expenses, esto pasa todas las expenses sin filtro, pero a mi me interesa filtrarlos
        expenses: selectExpenses(state.expenses, state.filter)
        //como habia seteado el filtro por palabra water, me va a salir un solo expense representado en expense list
    };
};
// y pasarle esto de argumento a connect

export default connect(mapStateToProps)(ExpenseList);

//esto se lo voy a pasar al dashboard