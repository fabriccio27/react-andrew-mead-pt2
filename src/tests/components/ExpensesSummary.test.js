import {shallow} from "enzyme";
import React from "react";
import {ExpensesSummary} from "../../components/ExpensesSummary";

test("should render correctly ExpensesSummary with no expenses", ()=>{
    const wrapper =  shallow(<ExpensesSummary/>);
    expect(wrapper).toMatchSnapshot();
});
test("should render correctly ExpensesSummary when passing expenses", ()=>{
    const wrapper = shallow(<ExpensesSummary expensesCount={3} expensesTotal={18999}/>);
    expect(wrapper).toMatchSnapshot();
});