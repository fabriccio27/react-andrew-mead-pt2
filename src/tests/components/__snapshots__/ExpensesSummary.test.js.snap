// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly ExpensesSummary when passing expenses 1`] = `
<div>
  <p>
    Total amount: 
    $189.99
     for 
    3
     expense(s)
  </p>
</div>
`;

exports[`should render correctly ExpensesSummary with no expenses 1`] = `
<div>
  <p>
    No expenses selected
  </p>
</div>
`;
